src/modules/csm/distribution.py

```diff
@@ -0,0 +1,222 @@
+            frame_blockstamp = self._get_frame_blockstamp(blockstamp, to_epoch)
+            frame_module_validators = self._get_module_validators(frame_blockstamp)
+
+            total_rewards_to_distribute = self.w3.csm.fee_distributor.shares_to_distribute(frame_blockstamp.block_hash)
+            rewards_to_distribute_in_frame = total_rewards_to_distribute - (total_distributed_rewards + total_rebate)
+
+            rewards_in_frame, log = self._calculate_distribution_in_frame(
+                frame, frame_blockstamp, rewards_to_distribute_in_frame, frame_module_validators
+            )
+
+            total_distributed_rewards += log.distributed_rewards
+            total_rebate += log.rebate_to_protocol
+
+            self.validate_distribution(total_distributed_rewards, total_rebate, total_rewards_to_distribute)
+
+            for no_id, rewards in rewards_in_frame.items():
+                total_rewards_map[no_id] += rewards
+
+            logs.append(log)
+
+        return total_distributed_rewards, total_rewards_map, total_rebate, logs
+
+    def _get_frame_blockstamp(self, blockstamp: ReferenceBlockStamp, to_epoch: EpochNumber) -> ReferenceBlockStamp:
+        if to_epoch != blockstamp.ref_epoch:
+            return self._get_ref_blockstamp_for_frame(blockstamp, to_epoch)
+        return blockstamp
+
+    def _get_ref_blockstamp_for_frame(
+        self, blockstamp: ReferenceBlockStamp, frame_ref_epoch: EpochNumber
+    ) -> ReferenceBlockStamp:
+        return get_reference_blockstamp(
+            cc=self.w3.cc,
+            ref_slot=self.converter.get_epoch_last_slot(frame_ref_epoch),
+            ref_epoch=frame_ref_epoch,
+            last_finalized_slot_number=blockstamp.slot_number,
+        )
+
+    def _get_module_validators(self, blockstamp: ReferenceBlockStamp) -> ValidatorsByNodeOperator:
+        return self.w3.lido_validators.get_module_validators_by_node_operators(
+            StakingModuleAddress(self.w3.csm.module.address), blockstamp
+        )
+
+    def _calculate_distribution_in_frame(
+        self,
+        frame: Frame,
+        blockstamp: ReferenceBlockStamp,
+        rewards_to_distribute: int,
+        operators_to_validators: ValidatorsByNodeOperator,
+    ) -> tuple[dict[NodeOperatorId, int], FramePerfLog]:
+        total_rebate_share = 0
+        participation_shares: defaultdict[NodeOperatorId, int] = defaultdict(int)
+        log = FramePerfLog(blockstamp, frame)
+
+        network_perf = self._get_network_performance(frame)
+
+        for (_, no_id), validators in operators_to_validators.items():
+            logger.info({"msg": f"Calculating distribution for {no_id=}"})
+            log_operator = log.operators[no_id]
+
+            curve_id = self.w3.csm.accounting.get_bond_curve_id(no_id, blockstamp.block_hash)
+            perf_coeffs, perf_leeway, reward_share = self._get_curve_params(curve_id, blockstamp)
+
+            sorted_validators = sorted(validators, key=lambda v: v.index)
```

Comment:
The `validators` variable includes exited validators. Is it as intended?

Reply:
Result of refactoring. There should be active only

Reply:
Fixed